Enhancement: detect and recover when feature branch is checked out in main worktree

## Problem

When \`feature/issue-N\` is checked out in the main worktree (e.g. left over from a previous builder run or manual checkout), the shepherd fails to create a new isolated worktree for the issue:

\`\`\`
fatal: 'feature/issue-2853' is already used by worktree at '/Users/rwalters/GitHub/loom'
ERROR: Failed to create worktree
\`\`\`

There is no detection, recovery, or actionable guidance — the shepherd simply fails and returns the issue to the queue.

## Expected Behavior

The shepherd (or \`worktree.sh\`) should detect this specific condition and either:

1. **Auto-recover**: switch the main worktree back to \`main\` (if it has no uncommitted changes) before creating the issue worktree
2. **Fail with guidance**: emit a clear, actionable message:
   \`\`\`
   [ERROR] Cannot create worktree for issue #2853: branch 'feature/issue-2853'
           is already checked out at '/Users/rwalters/GitHub/loom' (main worktree).
   
   To fix: cd /Users/rwalters/GitHub/loom && git checkout main
   \`\`\`

## Detection

The git error message contains \`is already used by worktree at\` — this pattern can be matched to distinguish this case from other worktree creation failures.

## References

- Observed during shepherd run for #2853 (2026-02-19)
- Related: #2918 (worktree conflict counted as systematic failure)
- \`worktree.sh\` and \`loom-tools/src/loom_tools/shepherd/phases/builder.py\`

--- a/defaults/scripts/worktree.sh
+++ b/defaults/scripts/worktree.sh
@@ -371,7 +371,152 @@ if [[ "$JSON_OUTPUT" != "true" ]]; then
     echo ""
 fi
 
-if git worktree add "${CREATE_ARGS[@]}"; then
+# Helper: attempt recovery when feature branch is checked out in the main worktree.
+# This happens when a previous builder manually checked out feature/issue-N in the
+# main workspace and left it there.  Git refuses to create a new worktree for that
+# branch: "fatal: 'feature/issue-N' is already used by worktree at '<main-path>'"
+#
+# Recovery strategy:
+#   1. Detect the "already used by worktree at" pattern in stderr
+#   2. Confirm the conflicting worktree is the main workspace (not a feature worktree)
+#   3. If main workspace is clean: auto-switch it back to main and retry
+#   4. If main workspace has uncommitted changes: emit an actionable error message
+_handle_feature_branch_in_main_worktree() {
+    local error_output="$1"
+    local branch="$2"
+
+    # Only act on the specific "already used by worktree at" error
+    if ! echo "$error_output" | grep -q "is already used by worktree at"; then
+        return 1  # Not this error — caller should fail normally
+    fi
+
+    # Extract the conflicting worktree path from the error message
+    # Example: "fatal: 'feature/issue-2853' is already used by worktree at '/Users/rwalters/GitHub/loom'"
+    local conflict_path
+    conflict_path=$(echo "$error_output" | grep -o "is already used by worktree at '[^']*'" | sed "s/is already used by worktree at '//;s/'$//")
+
+    if [[ -z "$conflict_path" ]]; then
+        # Could not parse path — emit a generic actionable message (human-readable only)
+        if [[ "$JSON_OUTPUT" != "true" ]]; then
+            print_error "Cannot create worktree: branch '$branch' is already checked out in another worktree."
+            echo ""
+            echo "  The branch is in use elsewhere. To free it, find the worktree with:"
+            echo "    git worktree list"
+            echo "  Then switch that worktree to main:"
+            echo "    cd <worktree-path> && git checkout main"
+        fi
+        return 0  # Handled (with human-readable message), no retry possible
+    fi
+
+    # Determine the main workspace path
+    local main_workspace
+    main_workspace=$(git rev-parse --git-common-dir 2>/dev/null)
+    main_workspace=$(dirname "$main_workspace" 2>/dev/null)
+
+    # Resolve both paths to absolute for comparison
+    local abs_conflict abs_main
+    abs_conflict=$(cd "$conflict_path" 2>/dev/null && pwd) || abs_conflict="$conflict_path"
+    abs_main=$(cd "$main_workspace" 2>/dev/null && pwd) || abs_main="$main_workspace"
+
+    if [[ "$abs_conflict" != "$abs_main" ]]; then
+        # Conflicting worktree is not the main workspace — it's a different issue worktree.
+        # This is unusual but can happen. Emit actionable guidance without auto-recovery.
+        if [[ "$JSON_OUTPUT" != "true" ]]; then
+            print_error "Cannot create worktree for branch '$branch':"
+            echo "  Branch is already checked out at: $conflict_path"
+            echo ""
+            echo "  To fix:"
+            echo "    cd $conflict_path && git checkout main"
+        fi
+        return 0  # Handled (with error message), no retry
+    fi
+
+    # The conflict is in the main workspace. Check for uncommitted changes.
+    local uncommitted
+    uncommitted=$(git -C "$abs_conflict" status --porcelain 2>/dev/null)
+
+    if [[ -n "$uncommitted" ]]; then
+        # Main workspace has uncommitted changes — cannot auto-recover safely
+        if [[ "$JSON_OUTPUT" != "true" ]]; then
+            print_error "Cannot create worktree for issue #$ISSUE_NUMBER: branch '$branch'"
+            echo "  is already checked out at '$abs_conflict' (main worktree)."
+            echo ""
+            echo "  The main worktree has uncommitted changes — cannot auto-switch."
+            echo "  To fix manually:"
+            echo "    cd $abs_conflict"
+            echo "    git stash  # or commit your changes"
+            echo "    git checkout main"
+            echo "  Then rerun: ./.loom/scripts/worktree.sh $ISSUE_NUMBER"
+        fi
+        return 0  # Handled (with error message), no retry
+    fi
+
+    # Main workspace is clean — auto-switch to main and signal caller to retry
+    if [[ "$JSON_OUTPUT" != "true" ]]; then
+        print_warning "Branch '$branch' is checked out in the main worktree."
+        print_info "Main worktree is clean — auto-switching to main branch..."
+    fi
+
+    if git -C "$abs_conflict" checkout main 2>/dev/null; then
+        if [[ "$JSON_OUTPUT" != "true" ]]; then
+            print_success "Main worktree switched to main branch"
+        fi
+        return 2  # Signal: auto-recovered, caller should retry
+    else
+        if [[ "$JSON_OUTPUT" != "true" ]]; then
+            print_error "Failed to switch main worktree to main branch."
+            echo "  To fix manually:"
+            echo "    cd $abs_conflict && git checkout main"
+            echo "  Then rerun: ./.loom/scripts/worktree.sh $ISSUE_NUMBER"
+        fi
+        return 0  # Handled (with error message), no retry
+    fi
+}
+
+_try_worktree_add() {
+    # Capture stderr separately so we can inspect it on failure while still
+    # showing stdout (git progress messages like "Preparing worktree...") to user.
+    local stderr_file
+    stderr_file=$(mktemp /tmp/loom-worktree-stderr-$$-XXXXXX)
+
+    git worktree add "${CREATE_ARGS[@]}" 2>"$stderr_file"
+    local exit_code=$?
+
+    if [[ $exit_code -eq 0 ]]; then
+        rm -f "$stderr_file"
+        return 0
+    fi
+
+    local worktree_error
+    worktree_error=$(cat "$stderr_file")
+    rm -f "$stderr_file"
+
+    # Attempt recovery for the "feature branch in main worktree" case.
+    # Wrap in a subshell result capture to safely handle non-zero returns
+    # without triggering set -e (we use exit code 2 as a retry signal).
+    local recovery_code=0
+    _handle_feature_branch_in_main_worktree "$worktree_error" "$BRANCH_NAME" && recovery_code=0 || recovery_code=$?
+
+    if [[ $recovery_code -eq 2 ]]; then
+        # Auto-recovered: retry worktree creation once
+        if [[ "$JSON_OUTPUT" != "true" ]]; then
+            print_info "Retrying worktree creation..."
+        fi
+        git worktree add "${CREATE_ARGS[@]}"
+        return $?
+    fi
+
+    if [[ $recovery_code -eq 1 ]]; then
+        # _handle_feature_branch_in_main_worktree returned 1 (not this error type)
+        # Print the original git error since nothing else has
+        echo "$worktree_error" >&2
+    fi
+    # recovery_code == 0 means error was handled and message already printed
+    return 1
+}
+
+
+if _try_worktree_add; then
     # Get absolute path to worktree
     ABS_WORKTREE_PATH=$(cd "$WORKTREE_PATH" && pwd)
 
@@ -520,8 +665,7 @@ if git worktree add "${CREATE_ARGS[@]}"; then
 else
     if [[ "$JSON_OUTPUT" == "true" ]]; then
         echo '{"success": false, "error": "Failed to create worktree"}'
-    else
-        print_error "Failed to create worktree"
     fi
+    # Human-readable error already printed by _try_worktree_add / _handle_feature_branch_in_main_worktree
     exit 1
 fi

--- a/loom-tools/src/loom_tools/worktree.py
+++ b/loom-tools/src/loom_tools/worktree.py
@@ -324,6 +324,118 @@ def check_worktree() -> int:
         return 1
 
 
+def _handle_feature_branch_in_main_worktree(
+    error_output: str,
+    branch_name: str,
+    issue_number: int,
+    json_output: bool = False,
+) -> tuple[bool, bool]:
+    """Detect and recover when a feature branch is checked out in the main worktree.
+
+    This condition arises when a previous builder manually checked out
+    ``feature/issue-N`` in the main workspace and left it there.  Git
+    refuses to create a new worktree for that branch with the error:
+    ``fatal: 'feature/issue-N' is already used by worktree at '<path>'``
+
+    Recovery strategy:
+    1. Detect the "already used by worktree at" pattern in stderr.
+    2. Confirm the conflicting worktree is the main workspace.
+    3. If main workspace is clean: auto-switch to ``main`` and return
+       ``(handled=True, retry=True)`` so the caller can retry.
+    4. If main workspace has uncommitted changes: emit an actionable
+       error and return ``(handled=True, retry=False)``.
+
+    Returns:
+        A (handled, retry) tuple:
+        - ``(False, False)`` — not this error, caller should propagate normally.
+        - ``(True, False)``  — this error, handled with message, no retry.
+        - ``(True, True)``   — this error, auto-recovered, caller should retry.
+    """
+    if "is already used by worktree at" not in error_output:
+        return False, False
+
+    # Extract the conflicting worktree path from the error message.
+    # Example: "fatal: 'feature/issue-2853' is already used by worktree at '/path'"
+    import re
+
+    match = re.search(r"is already used by worktree at '([^']+)'", error_output)
+    if not match:
+        # Could not parse path — emit actionable guidance and fail.
+        if not json_output:
+            log_error(
+                f"Cannot create worktree: branch '{branch_name}' is already "
+                "checked out in another worktree."
+            )
+            print()
+            print("  The branch is in use elsewhere. To free it, find the worktree with:")
+            print("    git worktree list")
+            print("  Then switch that worktree to main:")
+            print("    cd <worktree-path> && git checkout main")
+        return True, False
+
+    conflict_path = pathlib.Path(match.group(1)).resolve()
+
+    # Determine the main workspace path for comparison.
+    main_workspace = _get_main_workspace()
+    if main_workspace is None:
+        # Cannot determine main workspace — fall back to generic failure.
+        return False, False
+
+    abs_main = main_workspace.resolve()
+
+    if conflict_path != abs_main:
+        # Conflicting worktree is not the main workspace — it's another issue
+        # worktree.  Emit actionable guidance without auto-recovery.
+        if not json_output:
+            log_error(f"Cannot create worktree for branch '{branch_name}':")
+            print(f"  Branch is already checked out at: {conflict_path}")
+            print()
+            print("  To fix:")
+            print(f"    cd {conflict_path} && git checkout main")
+        return True, False
+
+    # The conflict is in the main workspace.  Check for uncommitted changes.
+    status_result = _run_git(["status", "--porcelain"], cwd=abs_main, check=False)
+    uncommitted = status_result.stdout.strip() if status_result.returncode == 0 else ""
+
+    if uncommitted:
+        # Main workspace has uncommitted changes — cannot auto-recover safely.
+        if not json_output:
+            log_error(
+                f"Cannot create worktree for issue #{issue_number}: "
+                f"branch '{branch_name}'"
+            )
+            print(
+                f"  is already checked out at '{abs_main}' (main worktree)."
+            )
+            print()
+            print("  The main worktree has uncommitted changes — cannot auto-switch.")
+            print("  To fix manually:")
+            print(f"    cd {abs_main}")
+            print("    git stash  # or commit your changes")
+            print("    git checkout main")
+            print(f"  Then rerun: ./.loom/scripts/worktree.sh {issue_number}")
+        return True, False
+
+    # Main workspace is clean — auto-switch to main.
+    if not json_output:
+        log_warning(f"Branch '{branch_name}' is checked out in the main worktree.")
+        log_info("Main worktree is clean — auto-switching to main branch...")
+
+    checkout_result = _run_git(["checkout", "main"], cwd=abs_main, check=False)
+    if checkout_result.returncode == 0:
+        if not json_output:
+            log_success("Main worktree switched to main branch")
+        return True, True  # Auto-recovered: retry worktree creation
+    else:
+        if not json_output:
+            log_error("Failed to switch main worktree to main branch.")
+            print("  To fix manually:")
+            print(f"    cd {abs_main} && git checkout main")
+            print(f"  Then rerun: ./.loom/scripts/worktree.sh {issue_number}")
+        return True, False
+
+
 def create_worktree(
     issue_number: int,
     custom_branch: str | None = None,
@@ -462,7 +574,32 @@ def create_worktree(
 
     result = _run_git(["worktree", "add"] + create_args, check=False)
     if result.returncode != 0:
-        return WorktreeResult(success=False, error="Failed to create worktree")
+        error_output = (result.stderr or result.stdout or "").strip()
+        handled, should_retry = _handle_feature_branch_in_main_worktree(
+            error_output, branch_name, issue_number, json_output
+        )
+
+        if should_retry:
+            # Auto-recovered (main worktree switched to main): retry once.
+            if not json_output:
+                log_info("Retrying worktree creation...")
+            result = _run_git(["worktree", "add"] + create_args, check=False)
+            if result.returncode != 0:
+                error_detail = (result.stderr or result.stdout or "").strip()
+                return WorktreeResult(
+                    success=False,
+                    error=f"Failed to create worktree after auto-recovery: {error_detail}",
+                )
+        elif handled:
+            # Error was handled with a message above — return clean failure.
+            return WorktreeResult(success=False, error="Failed to create worktree")
+        else:
+            # Unrecognised error — propagate git's message.
+            return WorktreeResult(
+                success=False,
+                error=f"Failed to create worktree: {error_output}" if error_output
+                else "Failed to create worktree",
+            )
 
     # Get absolute path
     abs_worktree_path = worktree_path.resolve()

--- a/loom-tools/tests/test_worktree.py
+++ b/loom-tools/tests/test_worktree.py
@@ -24,8 +24,12 @@ import pathlib
 
 import pytest
 
+import subprocess
+from unittest.mock import MagicMock, patch
+
 from loom_tools.worktree import (
     WorktreeResult,
+    _handle_feature_branch_in_main_worktree,
     main,
 )
 
@@ -193,3 +197,173 @@ class TestDocumentedBehaviorDifferences:
         and avoids needing to checkout or update the local main branch.
         """
         pass
+
+    def test_feature_branch_in_main_worktree_recovery(self) -> None:
+        """Document: worktree creation detects and recovers from feature branch in main worktree.
+
+        When git worktree add fails with "is already used by worktree at <main-path>",
+        the helper detects whether the conflict is in the main worktree and
+        auto-switches it to main (if clean) before retrying.
+
+        This fixes the failure observed in shepherd runs where a previous builder
+        left feature/issue-N checked out in the main workspace.
+        See issue #2924.
+        """
+        from loom_tools.worktree import _handle_feature_branch_in_main_worktree
+
+        assert callable(_handle_feature_branch_in_main_worktree)
+
+
+class TestHandleFeatureBranchInMainWorktree:
+    """Tests for the feature-branch-in-main-worktree detection and recovery helper."""
+
+    def test_non_matching_error_returns_not_handled(self) -> None:
+        """Unrelated git errors are not handled by this function."""
+        handled, retry = _handle_feature_branch_in_main_worktree(
+            "fatal: some other git error",
+            "feature/issue-42",
+            42,
+        )
+        assert not handled
+        assert not retry
+
+    def test_empty_error_returns_not_handled(self) -> None:
+        """Empty error output is not handled."""
+        handled, retry = _handle_feature_branch_in_main_worktree(
+            "",
+            "feature/issue-42",
+            42,
+        )
+        assert not handled
+        assert not retry
+
+    def test_matching_error_with_no_parseable_path_returns_handled(
+        self,
+        capsys: pytest.CaptureFixture[str],
+    ) -> None:
+        """Error with unparseable path is handled (message printed) but no retry."""
+        # Error message that matches but path extraction fails
+        error = "fatal: 'feature/issue-42' is already used by worktree at"
+        handled, retry = _handle_feature_branch_in_main_worktree(
+            error,
+            "feature/issue-42",
+            42,
+        )
+        assert handled
+        assert not retry
+        captured = capsys.readouterr()
+        # Should print guidance about git worktree list
+        assert "git worktree list" in captured.out or "git worktree list" in captured.err
+
+    def test_matching_error_conflict_not_main_workspace(
+        self,
+        tmp_path: pathlib.Path,
+        capsys: pytest.CaptureFixture[str],
+        monkeypatch: pytest.MonkeyPatch,
+    ) -> None:
+        """Error where conflict is NOT in the main workspace gets handled with guidance."""
+        conflict_path = tmp_path / "some-other-worktree"
+        conflict_path.mkdir()
+
+        # Simulate a path that doesn't match the main workspace
+        error = (
+            f"fatal: 'feature/issue-42' is already used by worktree at '{conflict_path}'"
+        )
+
+        # Mock _get_main_workspace to return a different path
+        main_path = tmp_path / "main-workspace"
+        main_path.mkdir()
+        with patch("loom_tools.worktree._get_main_workspace", return_value=main_path):
+            handled, retry = _handle_feature_branch_in_main_worktree(
+                error,
+                "feature/issue-42",
+                42,
+            )
+
+        assert handled
+        assert not retry
+        captured = capsys.readouterr()
+        # Should print guidance showing the conflict path
+        combined = captured.out + captured.err
+        assert str(conflict_path) in combined
+
+    def test_matching_error_main_workspace_with_uncommitted_changes(
+        self,
+        tmp_path: pathlib.Path,
+        capsys: pytest.CaptureFixture[str],
+    ) -> None:
+        """When main workspace has uncommitted changes, handled=True retry=False."""
+        main_path = tmp_path / "main-workspace"
+        main_path.mkdir()
+        conflict_path = main_path  # conflict IS the main workspace
+
+        error = (
+            f"fatal: 'feature/issue-42' is already used by worktree at '{conflict_path}'"
+        )
+
+        def mock_run_git(
+            args: list[str],
+            cwd: pathlib.Path | str | None = None,
+            check: bool = True,
+        ) -> subprocess.CompletedProcess[str]:
+            if args == ["status", "--porcelain"]:
+                result: subprocess.CompletedProcess[str] = MagicMock()
+                result.returncode = 0
+                result.stdout = "M some_file.py\n"
+                return result
+            # Fallback
+            result = MagicMock()
+            result.returncode = 0
+            result.stdout = ""
+            return result
+
+        with patch("loom_tools.worktree._get_main_workspace", return_value=main_path):
+            with patch("loom_tools.worktree._run_git", side_effect=mock_run_git):
+                handled, retry = _handle_feature_branch_in_main_worktree(
+                    error,
+                    "feature/issue-42",
+                    42,
+                )
+
+        assert handled
+        assert not retry
+        captured = capsys.readouterr()
+        combined = captured.out + captured.err
+        # Should mention uncommitted changes
+        assert "uncommitted" in combined.lower() or "stash" in combined.lower()
+
+    def test_matching_error_main_workspace_clean_auto_recovers(
+        self,
+        tmp_path: pathlib.Path,
+        capsys: pytest.CaptureFixture[str],
+    ) -> None:
+        """When main workspace is clean, auto-switch to main and signal retry."""
+        main_path = tmp_path / "main-workspace"
+        main_path.mkdir()
+        conflict_path = main_path
+
+        error = (
+            f"fatal: 'feature/issue-42' is already used by worktree at '{conflict_path}'"
+        )
+
+        def mock_run_git(
+            args: list[str],
+            cwd: pathlib.Path | str | None = None,
+            check: bool = True,
+        ) -> subprocess.CompletedProcess[str]:
+            result: MagicMock = MagicMock(spec=subprocess.CompletedProcess)
+            result.returncode = 0
+            result.stdout = ""
+            result.stderr = ""
+            return result  # type: ignore[return-value]
+
+        with patch("loom_tools.worktree._get_main_workspace", return_value=main_path):
+            with patch("loom_tools.worktree._run_git", side_effect=mock_run_git):
+                handled, retry = _handle_feature_branch_in_main_worktree(
+                    error,
+                    "feature/issue-42",
+                    42,
+                )
+
+        assert handled
+        assert retry  # Should signal caller to retry worktree creation

